localization second part: Localization ffg strings

eorahil opened the pull request:
* Corrected multilang string generation with commas and quotes.
* Translating all text at the begining of the scenario lasts too long. And after language change everything needed to be refreshed. I changed the system not to translate keys until screen displaying. Load time is much shorter and after language change nothing need to be refreshed.
* DictionaryI18n items loads the localization file raw data and create dictionary entrys as soon as they are needed. (unused keys wont become dict entrys).
* split IniRead class into IniRead and LocalizationRead in order to separate in:
  * IniRead: Converts a ini file into dictionary with entrys without translations (keys remains the same)
  * LocalizationRead: Loads dictionaries (ffg for now) and defines static methods to translate keys,
* Created the StringKey class to represent translatable key items (ffg ones for now). Changed string attributes of game components witch need to be translated.
* Slightly changed hero selection in order to avoid name conflicts.
* Created a ValkyrieDebug class to intercept log traces and disable all if needed in order to create individual unit tests.
* Renamed StringI18n to EntryI18n.
* Changed all strings that has an existing key in localization.txt file (hero names, monster names, expansion names).
* Corrected tiles related issues descripted in #79 and #80
* Fixed MoM character problem: There were 2 gloria goldbergs and no rita young.
* Merged development branch including puzzle adds

NPBruce commented:
I have finished changes for release 0.7.0 and am starting testing.  It would be great if you could help and report any issues you find.  I have cherry picked the tile name issues.  It is up to you if you want to get this pull request through now or wait for the next release, let me know.

NPBruce commented:
I have just fixed Rita now.  Yes the test branch is development.

eorahil commented:
I stated testing. I found that when you click "Start Scenario" or "Scenario edit" it lasts few seconds because its reading all localization.txt entries and processing it. In my pull request i fixed it by not processing all entries in localization file. Instead it gets the localization file into rawdata and gets localized text as soon as it need to be translated. It makes the app faster.
The problem is that if we pull the request, someone who selects a non english language will get ffg texts in foreign language and valkyrie and quest texts in english. T_T.

NPBruce commented:
I do like that change.  I can do something similar, but it would make your merge harder.  We could disable language selection for now.  If we are merging this there are a few other things to clean up.

eorahil commented:
ok, i will disable the options screen and then the language wont be selectable.

eorahil commented:
Ok. I merged all, fixed some problems, and removed the options menu. Now language cant be selected
i think it can be pulled correctly
I saw the House Linch adventure. Its amazing but, don't you think that converting a non public non free adventure (its included with mom1e) can carry problems with ffg?

NPBruce commented:
This is part of why I removed the included quests, if it does get taken down Valkyrie should be safe.  I figured I could do that easier than come up with a completely new scenario that showed off what can be done, plus it forced me to expand/fix Valkyrie rather than work with it's limitations.

NPBruce commented:
I have merged this to an 'integration' branch and fixed some of the issues there (not great fixes, but will do for now).

Will test a bit then merge from there to dev.

eorahil commented on Assets/Scripts/Content/ContentData.cs:
```diff
@@ -482,7 +483,7 @@ void AddContent(string name, Dictionary<string, string> content, string path, st
             PerilData d = new PerilData(name, content);
             // Ignore invalid entry
-            if (d.name.Equals(""))
+            if (d.sectionName.Equals(""))
```
sorry about that. The main problem with name is that it sometimes is used as a localizable text (public name) ane others like a code (internal name). But all of them inherits from the GenericData class and I figured out changing name to StringKey and when name is useda as inernal code, use sectionName instead. On the investigator selection dialog it worked fine but in peril looks like there is some problem? 
I thought about spliting name into codeName and publicName but I wasn't sure about where to use one or other.

NPBruce replied:
I don't think this test is useful anyway for perils.  Will review later.

eorahil commented on Assets/Scripts/Quest/QuestMonster.cs:
```diff
@@ -46,8 +49,8 @@ public QuestMonster(QuestData.UniqueMonster qm) : base()
         }
 
         // Read info from quest data or base type
-        info = EventManager.SymbolReplace(qm.info);
```
I think the only place to make symbol replace is from a visual component (textBox) to internal data. After this point, there is no need to replace again because FFG font has all this symbols inside.
After translation, my first idea is to create a textbox rich editor with buttons to insert these special ymbols directly without transformations. and a transform button to make all transformations and look how it will look like before saving.

NPBruce replied:
But here it is a monster type defined by the quest, so it does need symbol replace.

I would like the quest translations at least (if not valkyrie translations) to use {symbol} notation on disk because it is readable in a text editor.

eorahil replied:
Ok. I will take this in consideration. Then the default symbol store should be the {name} notation at least for quests.
I will solve it in next localization commit.
